internal/exec/stages/files/units.go

```diff
@@ -64,10 +65,77 @@ func (s *stage) createUnits(config types.Config) error {
 			s.relabel(relabelpath)
 		}
 	}
-	// and relabel the preset file itself if we enabled/disabled something
-	if enabledOneUnit {
-		s.relabel(util.PresetPath)
+	if err := s.createSystemdUnits(config); err != nil {
+		return err
+	}
+	return nil
+}
+
+// createSystemdUnits creates the enabled and disabled units listed under
+// systemd units. This function is primarily written to support instantiated
+// units to be enabled via presets.For more information: https://github.com/systemd/systemd/pull/9901/files
+func (s *stage) createSystemdUnits(config types.Config) error {
+	var instantiatedEnableServices strings.Builder
+	if isEnableUnit {
+		for _, enableUnit := range enabledUnits {
+			if strings.Contains(enableUnit.Name, "@") {
+				at := strings.Index(enableUnit.Name, "@")
+				dot := strings.Index(enableUnit.Name, ".")
+				instance := enableUnit.Name[at+1 : dot]
+				serviceInstance := enableUnit.Name[0:at+1] + enableUnit.Name[dot:len(enableUnit.Name)]
```

Comment:
Thanks for the input. I'm trying to extract the word in between `@` and  `.` and not after `.` so, here it makes sense to include a slice index to be the end of the string.
For e.g. :
```
test: echo@foo.service
o/p: foo
```

Reply:
That's not what is being done on the highlighted line creating `serviceInstance` though (that's what you're doing for just `instance`). In your example if `echo@foo.service` is the value of `enableUnit.Name` then `serviceInstance` would equal `echo@.service` Here's a playground of those lines isolated: https://play.golang.org/p/OB_hMOfoCLL

Reply:
I think I might have confused you earlier. Here that logic is handling a feature present in this (https://github.com/systemd/systemd/pull/9901/files) PR where you can expect multiple instantiated units in a given ignition file, and our task is to append those services in a single line in the `20-ignition.preset` file.
For e.g.:
```
systemd:
  units:
    - name: echo@.service
        contents: |
          [Unit]
          Description=f
          [Service]
          Type=oneshot
          ExecStart=/bin/echo %i
          RemainAfterExit=yes
          [Install]
          WantedBy=multi-user.target
    - name: chronyd.service
      enabled: false
    - name: sshd.service
      enabled: true
    - name: echo@bar.service
      enabled: true
    - name: zincati.service
      enabled: false
    - name: echo@foo.service
      enabled: true
    - name: getty@tty1.service
      enabled: false
```
your `20-ignition.preset` file will have the following entries:
```
disable chronyd.service
enable sshd.service
enable foo.service
disable zincati.service
enable echo@.service foo bar
disable getty@.service tty1
```
So that logic is handling the above test case.